### ISSUE TITLE
[Bug] a2a3 FP32 matmul with b_trans=True gives wrong results for [16,16384] x [32,16384].T

### ISSUE THREAD

**Issue by high-cloud** [hw-native-sys/pypto#1213]
## Component

Backend / Codegen

## Description

A minimal FP32 `pl.matmul(..., b_trans=True)` reproducer with DeepSeek HC pre projection shape produces incorrect results on real a2a3 hardware. This does not involve padding or an intermediate assembled padded weight tensor.

The failing computation is:

```python
out[16, 32] = x[16, 16384] @ w[32, 16384].T
```

The program chunks K by 512 and uses `pl.matmul` for the first chunk plus `pl.matmul_acc` for the remaining chunks. The run completes successfully and only fails during golden validation, so this is a silent wrong-result issue.

This is related to #1212, but the trigger here is different: #1212 focuses on a column-strided lhs slice and does not use `b_trans`; this reproducer has no padding and uses the RHS transposed view generated by `b_trans=True`.

## Steps to Reproduce

Create `debug_hc_f32_matmul_nopad.py` with the following program and run it on a2a3:

```python
import pypto.language as pl

B = 16
S = 1
D = 4096
HC_MULT = 4
MIX_PAD = 32
HC_DIM = HC_MULT * D
T = B * S
K_CHUNK = 512
HC_DIM_BLOCKS = HC_DIM // K_CHUNK


def build_debug_hc_f32_matmul_nopad_program():
    @pl.program
    class DebugHcF32MatmulNoPad:
        @pl.function(type=pl.FunctionType.Opaque)
        def debug_hc_f32_matmul_nopad(
            self,
            x:   pl.Tensor[[T, HC_DIM],       pl.FP32],
            w:   pl.Tensor[[MIX_PAD, HC_DIM], pl.FP32],
            out: pl.Out[pl.Tensor[[T, MIX_PAD], pl.FP32]],
        ):
            with pl.at(level=pl.Level.CORE_GROUP, optimization=pl.chunked_loop_optimizer, name_hint="linear"):
                x_chunk0 = pl.slice(x, [T, K_CHUNK], [0, 0])
                w_chunk0 = pl.slice(w, [MIX_PAD, K_CHUNK], [0, 0])
                acc = pl.matmul(x_chunk0, w_chunk0, b_trans=True, out_dtype=pl.FP32)
                for kb in pl.range(1, HC_DIM_BLOCKS):
                    k0 = kb * K_CHUNK
                    x_chunk = pl.slice(x, [T, K_CHUNK], [0, k0])
                    w_chunk = pl.slice(w, [MIX_PAD, K_CHUNK], [0, k0])
                    acc = pl.matmul_acc(acc, x_chunk, w_chunk, b_trans=True)
                out = pl.assemble(out, acc, [0, 0])

    return DebugHcF32MatmulNoPad


def build_tensor_specs():
    import torch
    from golden import TensorSpec

    def init_x():
        return torch.rand(T, HC_DIM, dtype=torch.float32) - 0.5

    def init_w():
        return torch.randn(MIX_PAD, HC_DIM, dtype=torch.float32) / (HC_DIM ** 0.5)

    return [
        TensorSpec("x",   [T, HC_DIM],       torch.float32, init_value=init_x),
        TensorSpec("w",   [MIX_PAD, HC_DIM], torch.float32, init_value=init_w),
        TensorSpec("out", [T, MIX_PAD],      torch.float32, is_output=True),
    ]


def golden_debug_hc_f32_matmul_nopad(tensors):
    tensors["out"][:] = tensors["x"] @ tensors["w"].T


if __name__ == "__main__":
    import argparse
    from golden import RunConfig, run

    parser = argparse.ArgumentParser()
    parser.add_argument("-p", "--platform", type=str, default="a2a3",
                        choices=["a2a3", "a2a3sim", "a5", "a5sim"])
    parser.add_argument("-d", "--device", type=int, default=0)
    parser.add_argument("--runtime-profiling", action="store_true", default=False)
    args = parser.parse_args()

    result = run(
        program=build_debug_hc_f32_matmul_nopad_program(),
        tensor_specs=build_tensor_specs(),
        golden_fn=golden_debug_hc_f32_matmul_nopad,
        config=RunConfig(
            rtol=4e-3,
            atol=4e-3,
            compile=dict(dump_passes=True),
            runtime=dict(
                platform=args.platform,
                device_id=args.device,
                runtime_profiling=args.runtime_profiling,
            ),
        ),
    )
    if not result.passed:
        if result.error:
            print(result.error)
        raise SystemExit(1)
```

Command used locally from `pypto-lib`:

```bash
task-submit --device auto --max-time 0 --timeout 0 \
  --env PYTHONPATH --env ASCEND_HOME_PATH --env PTOAS_ROOT --env PTO_ISA_ROOT --env LD_LIBRARY_PATH \
  --run "python examples/models/deepseek/v4/debug_hc_f32_matmul_nopad.py -p a2a3 -d {}"
```

## Expected Behavior

The device output should match the torch golden:

```python
tensors["out"][:] = tensors["x"] @ tensors["w"].T
```

within `rtol=4e-3, atol=4e-3`.

## Actual Behavior

The run completes successfully but validation fails:

```text
[RUN]   'out' FAIL  shape=(16, 32) dtype=torch.float32
Output(s) does not match golden: ['out']
  'out' FAIL  shape=(16, 32) dtype=torch.float32
    Mismatched elements: 469/512  rtol=0.004 atol=0.004
    first 20 mismatches:
    [1] actual=0.10638493299484253, expected=0.027910670265555382
    [2] actual=-0.1785854548215866, expected=-0.18520885705947876
    [3] actual=-0.20775386691093445, expected=-0.20227229595184326
    [4] actual=-0.7271235585212708, expected=-0.6267380118370056
    [5] actual=0.5200693607330322, expected=0.6342579126358032
    [6] actual=0.08446823805570602, expected=0.15745769441127777
    [7] actual=-0.002614065306261182, expected=-0.06819099932909012
    [8] actual=-0.07401550561189651, expected=-0.1282496452331543
    [9] actual=-0.38961055874824524, expected=-0.4512384831905365
    [10] actual=0.2831008732318878, expected=0.28954562544822693
    [11] actual=0.5758091807365417, expected=0.6591705679893494
    [12] actual=0.31623518466949463, expected=0.45396795868873596
    [13] actual=-0.026538299396634102, expected=0.045861855149269104
    [14] actual=0.3246922791004181, expected=0.34157857298851013
    [15] actual=0.3566913902759552, expected=0.37198013067245483
    [16] actual=-0.09938869625329971, expected=-0.2057902216911316
    [17] actual=0.019643908366560936, expected=-0.06046643480658531
    [18] actual=-0.09270231425762177, expected=-0.09916370362043381
    [19] actual=0.42915284633636475, expected=0.49323421716690063
    [20] actual=0.16750812530517578, expected=0.12517079710960388
```

The corresponding generated PTO shows the RHS transposed view lowering to a DN layout:

```text
%w_chunk0__ssa_v0_view = pto.make_tensor_view %arg1,
  shape = [512, 32], strides = [1, 512] {layout = #pto.layout<dn>}

%w__ssa_v0_view = pto.make_tensor_view %arg3,
  shape = [16384, 32], strides = [1, 16384] {layout = #pto.layout<dn>}

pto.tmatmul ...
pto.tmatmul.acc ...
```

## Git Commit ID

`d4774abead4fadf09bafdf07ae3b104902c6e243`

## NPU Kind

Ascend 910B

## Host Platform

Linux (aarch64)

## Additional Context

Environment used for the repro:

| Component | Version |
| --- | --- |
| pypto | `d4774abead4fadf09bafdf07ae3b104902c6e243` (`main`) |
| pypto-lib | `d55031b` (`codex/reuse-claude-skills`, local repro script added) |
| simpler / pypto runtime | `16311c4d8011f88032b8ec2391ec2d0051d977ee` (detached HEAD) |
| ptoas | `0.33` |
| CANN | `8.5.0.alpha001` |
| Python | `3.10.20` |

I also tried a padded variant that first pads `hc_fn[24, 16384]` to `[32, 16384]` and then runs the same matmul shape. It fails similarly (`475/512` mismatches), while this nopad version fails with `469/512` mismatches. That suggests the pad/assemble step is not the primary trigger; the issue is reproducible with a direct `[32, 16384]` RHS tensor.

**Comment by high-cloud** [comment]
Additional data point: the same nopad shape also fails with BF16 inputs and FP32 output/accumulation.

Minimal variant:

```python
x:   pl.Tensor[[16, 16384], pl.BF16]
w:   pl.Tensor[[32, 16384], pl.BF16]
out: pl.Out[pl.Tensor[[16, 32], pl.FP32]]

acc = pl.matmul(x[:, :512], w[:, :512], b_trans=True, out_dtype=pl.FP32)
for kb in pl.range(1, 32):
    acc = pl.matmul_acc(acc, x[:, kb*512:(kb+1)*512], w[:, kb*512:(kb+1)*512], b_trans=True)
```

golden:

```python
tensors["out"][:] = tensors["x"].float() @ tensors["w"].float().T
```

Run result on a2a3:

```text
[RUN]   'out' FAIL  shape=(16, 32) dtype=torch.float32
Output(s) does not match golden: ['out']
  'out' FAIL  shape=(16, 32) dtype=torch.float32
    Mismatched elements: 470/512  rtol=0.004 atol=0.004
    first 20 mismatches:
    [1] actual=-0.05008309707045555, expected=-0.01725451834499836
    [2] actual=0.3494800925254822, expected=0.40814706683158875
    [3] actual=-0.07202184200286865, expected=-0.08804616332054138
    [4] actual=-0.1786631941795349, expected=-0.08320370316505432
    [5] actual=-0.3155413866043091, expected=-0.3433859944343567
    [6] actual=0.48645633459091187, expected=0.4131370186805725
    [7] actual=0.22117218375205994, expected=0.30027490854263306
    [8] actual=-0.03553403913974762, expected=-0.12801045179367065
    [9] actual=0.2524164915084839, expected=0.23475222289562225
    [10] actual=0.2073739767074585, expected=0.228226900100708
    [11] actual=0.09012030065059662, expected=0.0841730609536171
    [12] actual=0.18979762494564056, expected=0.13485154509544373
    [13] actual=0.18292461335659027, expected=0.07901205122470856
    [14] actual=0.0700453594326973, expected=0.064936563372612
    [15] actual=-0.2621590495109558, expected=-0.2423160821199417
    [16] actual=0.365803599357605, expected=0.3499835133552551
    [17] actual=-0.036520786583423615, expected=-0.07987809926271439
    [18] actual=-0.12076500058174133, expected=-0.148794025182724
    [19] actual=-0.1258053183555603, expected=-0.1366456151008606
    [20] actual=-0.06292402744293213, expected=-0.11797817051410675
```

The generated PTO still uses a DN RHS view and cube matmul path:

```text
%w_chunk0__ssa_v0_view = pto.make_tensor_view %arg1, shape = [512, 32], strides = [1, 512] {layout = #pto.layout<dn>}: !pto.tensor_view<?x?xbf16>
%w__ssa_v0_view = pto.make_tensor_view %arg3, shape = [16384, 32], strides = [1, 16384] {layout = #pto.layout<dn>}: !pto.tensor_view<?x?xbf16>
pto.tmatmul ... dtype=bf16 ... outs dtype=f32
pto.tmatmul.acc ... dtype=bf16 ... outs dtype=f32
```

So this is not FP32-only; BF16 input + FP32 accumulation has the same wrong-result pattern for this `b_trans=True` RHS-DN path.

**Comment by high-cloud** [comment]
Additional K sweep for the FP32 `b_trans=True` nopad reproducer:

All cases compute:

```text
out[16, 32] = x[16, K] @ w[32, K].T
```

with `K_CHUNK_MAX=512`, so K=128/512 use a single `pl.matmul`, while K>=2048 use `pl.matmul + pl.matmul_acc`.

| K | K blocks | Result on a2a3 |
| --- | ---: | --- |
| 128 | 1 | PASS |
| 512 | 1 | PASS |
| 2048 | 4 | FAIL, 485/512 mismatches |
| 4096 | 8 | FAIL, 485/512 mismatches |
| 16384 | 32 | FAIL, 475/512 mismatches |

This suggests the trigger is not simply `b_trans=True` or a DN RHS view by itself. The wrong-result behavior starts once the `b_trans=True` RHS-DN path is used together with multi-block K accumulation (`matmul_acc`).

**Comment by high-cloud** [comment]
One more point in the K sweep: `K=1024` also fails.

| K | K blocks (`K_CHUNK=512`) | Result on a2a3 |
| --- | ---: | --- |
| 128 | 1 | PASS |
| 512 | 1 | PASS |
| 1024 | 2 | FAIL, 492/512 mismatches |
| 2048 | 4 | FAIL, 485/512 mismatches |
| 4096 | 8 | FAIL, 485/512 mismatches |
| 16384 | 32 | FAIL, 475/512 mismatches |

So the failure starts as soon as there is at least one `pl.matmul_acc(..., b_trans=True)` after the initial `pl.matmul(..., b_trans=True)`.

**Comment by high-cloud** [comment]
Retested after updating pypto to `42efdf7a` (`fix(codegen): Reuse matmul acc buffer`) on branch `issue-1213-matmul-acc-buffer`.

The generated PTO now does reuse the same accumulator tile in `tmatmul.acc`:

```text
%acc__tile = pto.alloc_tile ... loc=acc ...
pto.tmatmul ... outs(%acc__tile ...)
...
pto.tmatmul.acc ins(%acc__tile, %x_chunk__tile_Left, %w_chunk__tile_Right ...) outs(%acc__tile ...)
pto.tstore ins(%acc__tile ...)
```

However, the reproducer still fails on a2a3:

| K | Result |
| --- | --- |
| 1024 | FAIL, 484/512 mismatches |
| 16384 | FAIL, 471/512 mismatches |

So the buffer-reuse codegen shape is now fixed, but the wrong-result issue is not fully resolved yet.
